File: opt_graph.py
#!python
# -*- coding: utf-8 -*-
# Copyright (c) 2014, Sven Thiele <sthiele78@gmail.com>
#
# This file is part of iggy.
#
# iggy is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# iggy is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNEOS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with iggy.  If not, see <http://www.gnu.org/licenses/>.

from pyasp.asp import *
import argparse
from __iggy__ import query, utils, parsers

if __name__ == '__main__':
  desc = (
    'Opt-graph confronts a biological network given as interaction graphs with '
    'sets of experimental observations given as signs that represent the '
    'concentration changes between two measured states. Opt-graph computes the '
    'networks fitting the observation data by removing (or adding) a minimal '
    'number of edges in the given network')
  parser = argparse.ArgumentParser(description=desc)
  parser.add_argument('networkfile',
                      help='influence graph in SIF format')
  parser.add_argument('observationfiles',
                      help='directory of observations in bioquali format')

  parser.add_argument('--no_fwd_propagation',
    help="turn forward propagation OFF, default is ON",
    action="store_true")


  parser.add_argument('--no_founded_constraints',
    help='turn constraints OFF that every variation must be founded in an '
         'input, default is ON',
    action="store_true")

  parser.add_argument('--elempath',
    help=" a change must be explained by an elementary path from an input.",
    action="store_true")

  parser.add_argument('--depmat',
    help="combines multiple states, a change must be explained by an elementary path from an input.",
    action="store_true")    


  parser.add_argument('--autoinputs',
    help='compute possible inputs of the network (nodes with indegree 0)',
    action='store_true')

  parser.add_argument('--show_repairs',type=int, default=-1,
    help="number of repairs to show, default is OFF, 0=all")

  parser.add_argument('--repair_mode', type=int, default=1,
    help='choose repair mode:\n'
         ' 1 = add edges (default),\n'
         ' 2 = add + remove edges (opt-graph),\n'
         ' 3 = flip edges')


  args = parser.parse_args()

  net_string = args.networkfile
  obs_dir = args.observationfiles

  FP  = not (args.no_fwd_propagation)
  FC  = not (args.no_founded_constraints)
  EP  = args.elempath
  DM  = args.depmat
  print('_____________________________________________________________________\n')
  if DM :
    print(' + DepMat combines multiple states.')
    print(' + An elementary path from an input must exist to explain changes.')
    OS = False
    EP = True
    FP = True
    FC = True

  else :
    print(' + All observed changes must be explained by an predecessor.')
    OS = True
    if FP : print(' + 0-change must be explained.')
    if FC : print(' + All observed changes must be explained by an input.')
    if EP : print(' + An elementary path from an input must exist top explain changes.')

  print('_____________________________________________________________________')

  print('\nReading network',net_string, '... ',end='')
  net = parsers.readSIFGraph(net_string)
  print('done.')

  # gather stats on the network
  activations = set()
  inhibitions = set()
  nodes       = set()
  for a in net:
    if a.pred() == 'obs_elabel' :
      if a.arg(2) == '1'  : activations.add((a.arg(0),a.arg(1)))
      if a.arg(2) == '-1' : inhibitions.add((a.arg(0),a.arg(1)))
    if a.pred() == 'vertex' : nodes.add(a.arg(0))
  unspecified = activations & inhibitions

  print('\nNetwork stats:')
  print("         Nodes =", len(nodes))
  print("   Activations =", len(activations))
  print("   Inhibitions =", len(inhibitions))
  print("          Dual =", len(unspecified))


  flist = os.listdir(obs_dir)
  print('\nReading',len(flist),'observation sets from',obs_dir,'... ',end='')
  MU = TermSet()
  for f in flist :
    exp = os.path.join(obs_dir,f)
    mu  = parsers.readProfile(exp)
    MU  = TermSet(MU.union(mu))
  print('done.')

  print('\nChecking observations ... ',end='')
  contradictions = query.get_contradictory_obs(MU)
  print('done.')
  if len(contradictions) == 0 : print('\nObservations are OK.')
  else:
    print('\nContradictory observations found. Please correct manually.')
    for c in contradictions : print('  ',c)
    utils.clean_up()
    exit()

  if args.autoinputs :
    print('\nComputing input nodes ... ',end='')
    inputs = query.guess_inputs(net)
    net    = TermSet(net.union(inputs))
    print('done.')
    print('\nNumber of inputs =', len(inputs))

  net_with_data = TermSet(net.union(MU))

  if args.repair_mode==3 :
    print('\nComputing minimal number of flipped edges ... ',end='')
    (scenfit,repairs) = query.get_opt_flip_edges(net_with_data, OS, FP, FC, EP)
    print('done.')
    print('\nThe network and data can reach a scenfit of',scenfit,'with',repairs,'flipped edges.')

    if args.show_repairs >= 0 and repairs > 0:
      print('\nCompute optimal repairs ... ',end='')
      repairs = query.get_opt_repairs_flip_edges(net_with_data,args.show_repairs, OS, FP, FC, EP)
      print('done.')
      count=0
      for r in repairs :
        count += 1
        print('\nRepair ',str(count),':',sep='')
        utils.print_repairs(r)

  elif args.repair_mode==2 :
    print('\nComputing minimal number of changes add/remove edges ... ',end='')
    if EP :
      print('\n   using greedy method ... ',end='')
      (scenfit,redges) = query.get_opt_add_remove_edges_greedy(net_with_data)

      print('done.')
      print('\nThe network and data can reach a scenfit of',scenfit)
      #      ,'with', repairs,'removals and ',len(edges),'additions.')

      count_repairs = 0

      if args.show_repairs >= 0 :
        print('\nCompute optimal repairs ... ')
        print('   use greedily added edges ...')
        for (edges,repairs) in redges:
          if repairs > 0:
            repairs = query.get_opt_repairs_add_remove_edges_greedy(net_with_data,args.show_repairs,edges)

            for r in repairs :
              count_repairs+=1
              print('\nRepair ',str(count_repairs),':',sep='')
              for e in edges:
                print('   addedge',str(e)[10:],sep='')
              utils.print_repairs(r)
          else:
            count_repairs+=1
            print('\nRepair ',count_repairs,':',sep='')
            for e in edges:
              print('   addedge',str(e)[10:],sep='')

    else :
      (scenfit,repairscore) = query.get_opt_add_remove_edges(net_with_data, OS, FP, FC, EP)
      print('done.')
      print('\nThe network and data can reach a scenfit of',scenfit,'with repairs of score',str(repairscore)+'.')

      if args.show_repairs >= 0 and repairscore > 0:
        print('\nCompute optimal repairs ... ',end='')
        repairs = query.get_opt_repairs_add_remove_edges(net_with_data,args.show_repairs, OS, FP, FC, EP)
        print('done.')
        count = 0
        for r in repairs :
          count += 1
          print('\nRepair ',str(count),':',sep='')
          utils.print_repairs(r)

  else: # repair_mode==1
    print('\nComputing minimal number of removed edges ... ',end='')
    (scenfit,repairs) = query.get_opt_remove_edges(net_with_data, OS, FP, FC, EP)
    print('done.')
    print('\nThe network and data can reach a scenfit of',scenfit,'with',repairs,'removed edges.')

    if args.show_repairs >= 0 and repairs > 0:
      print('\nCompute optimal repairs ... ',end='')
      repairs = query.get_opt_repairs_remove_edges(net_with_data,args.show_repairs, OS, FP, FC, EP)
      print('done.')
      count=0
      for r in repairs :
        count += 1
        print('\nRepair ',str(count),':',sep='')
        utils.print_repairs(r)

utils.clean_up()

File: src/bin/optgraph.rs
use anyhow::{Context, Result};
use clap::Parser;
use clingo::FactBase;
use iggy::{cif_parser::Graph, misc::*, CheckResult::Inconsistent, *};
use itertools::Itertools;
use log::{error, info};
use std::{fmt, fs, fs::File, io::Write, path::PathBuf, str::FromStr};
use stderrlog;
use thiserror::Error;

/// Optgraph confronts interaction graph models with observations of (signed) changes between
/// two measured states.
/// Opt-graph computes networks fitting the observation data by removing (or adding) a minimal
/// number of edges in the given network.
#[derive(Parser, Debug)]
#[command(name = "optgraph", version, author)]
struct Opt {
    /// Influence graph in CIF format
    #[arg(short = 'n', long = "network", value_name = "FILE")]
    network_file: PathBuf,

    /// Directory of observations in bioquali format
    #[arg(short = 'o', long = "observations", value_name = "DIR")]
    observations_dir: PathBuf,

    /// Disable forward propagation constraints
    #[arg(long, conflicts_with = "depmat")]
    fwd_propagation_off: bool,

    /// Disable foundedness constraints
    #[arg(long, conflicts_with = "depmat")]
    founded_constraints_off: bool,

    /// Every change must be explained by an elementary path from an input
    #[arg(long)]
    elempath: bool,

    /// Combine multiple states, a change must be explained by an elementary path from an input
    #[arg(long)]
    depmat: bool,

    /// Declare nodes with indegree 0 as inputs
    #[arg(short = 'a', long)]
    auto_inputs: bool,

    /// Show N repairs, default is OFF, 0=all
    #[arg(short = 'r', long = "show-repairs", value_name = "N")]
    max_repairs: Option<u32>,

    /// REPAIR_MODE: remove = remove edges (default),
    ///              optgraph = add + remove edges,
    ///              flip = flip direction of edges
    #[arg(short = 'm', long)]
    repair_mode: Option<RepairMode>,

    /// Multithreading
    #[arg(short = 't', long, value_name = "N", default_value_t = 1)]
    threads: u8,

    /// Print JSON output
    #[arg(long)]
    json: bool,
}

#[derive(Debug, Error)]
struct InconsistenObsErr {
    inconsistencies: Vec<InconsistentObs>,
}
impl fmt::Display for InconsistenObsErr {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "Inconsistencies in observations ")?;
        for i in &self.inconsistencies {
            writeln!(f, "{i}")?;
        }
        Ok(())
    }
}
#[derive(Debug)]
struct InconsistentObs {
    observations_file: String,
    inconsistencies: Vec<String>,
}
impl fmt::Display for InconsistentObs {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(
            f,
            "Inconsistencies in observations file {}",
            self.observations_file
        )?;
        for i in &self.inconsistencies {
            writeln!(f, "- {i}")?;
        }
        Ok(())
    }
}
#[derive(Debug, Clone)]
enum RepairMode {
    Remove,
    OptGraph,
    Flip,
}
#[derive(Debug, Error)]
#[error("ParseRepairModeError: failed to parse repair mode. Possible values are: add, optgraph and flip. ")]
pub struct ParseRepairModeError;

impl FromStr for RepairMode {
    type Err = ParseRepairModeError;
    fn from_str(mode: &str) -> Result<Self, Self::Err> {
        match mode {
            "remove" => Ok(RepairMode::Remove),
            "optgraph" => Ok(RepairMode::OptGraph),
            "flip" => Ok(RepairMode::Flip),
            _ => Err(ParseRepairModeError),
        }
    }
}
fn main() {
    stderrlog::new()
        .module(module_path!())
        .verbosity(2)
        .init()
        .unwrap();
    if let Err(err) = run() {
        println!("\n## Error\n\n{err}");
        error!("{:?}", err);
        std::process::exit(1);
    }
}
fn run() -> Result<()> {
    let opt = Opt::parse();
    let mut out = std::io::stdout();
    if opt.json {
        writeln!(out, "{{")?;
    } else {
        writeln!(out, "# Optgraph Report")?;
    }

    let graph = read_network(&opt, &mut out)?;
    let graph_facts = graph.to_facts();

    let profiles = read_observations(&opt, &mut out)?;

    let setting = get_setting(&opt);
    if opt.json {
        write_setting_json(&mut out, &setting)?;
    } else {
        write_setting_md(&mut out, &setting)?;
    }

    let auto_inputs = compute_auto_inputs(&opt, &graph_facts, &mut out)?;

    if !opt.json {
        writeln!(out, "\n## Consistency results\n")?;
    }
    // Compute opt scenfit repair scores
    let (scenfit, repair_score, redges) = match opt.repair_mode {
        Some(RepairMode::OptGraph) if setting.ep => {
            info!("Computing repair through add/removing edges ... ");
            info!("using greedy method ... ");
            let (scenfit, repair_score, redges) = get_opt_add_remove_edges_greedy(
                &graph_facts,
                &profiles,
                &auto_inputs,
                opt.threads,
            )?;
            if opt.json {
                writeln!(out, ",\"scenfit\":{scenfit}")?;
                writeln!(out, ",\"repair score\":{repair_score}")?;
            } else {
                writeln!(
                    out,
                    "The network and data can reach a scenfit of {scenfit}."
                )?;
            }
            (scenfit, repair_score, redges)
            //   with {} removals and {} additions.", repairs, edges.len());
        }
        Some(RepairMode::OptGraph) if !setting.ep => {
            info!("Computing repair through add/removing edges ... ");
            let (scenfit, repair_score) =
                get_opt_add_remove_edges(&graph_facts, &profiles, &auto_inputs, &setting)?;
            if opt.json {
                writeln!(out, ",\"scenfit\":{scenfit}")?;
                writeln!(out, ",\"repair score\":{repair_score}")?;
            } else {
                writeln!(out, "The network and data can reach a scenfit of {scenfit} with repairs of score {repair_score}")?;
            }
            (scenfit, repair_score, vec![])
        }
        Some(RepairMode::Flip) => {
            info!("Computing repair through flipping edges ... ");
            let (scenfit, repair_score) =
                get_opt_flip_edges(&graph_facts, &profiles, &auto_inputs, &setting)?;
            if opt.json {
                writeln!(out, ",\"scenfit\":{scenfit}")?;
                writeln!(out, ",\"repair score\":{repair_score}")?;
            } else {
                writeln!(out, "The network and data can reach a scenfit of {scenfit} with {repair_score} flipped edges")?;
            }
            (scenfit, repair_score, vec![])
        }
        _ => {
            info!("Computing repair through removing edges ... ");
            let (scenfit, repair_score) =
                get_opt_remove_edges(&graph_facts, &profiles, &auto_inputs, &setting)?;
            if opt.json {
                writeln!(out, ",\"scenfit\":{scenfit}")?;
                writeln!(out, ",\"repair score\":{repair_score}")?;
            } else {
                writeln!(out,
                    "The network and data can reach a scenfit of {scenfit} with {repair_score} removed edges."
                )?;
            }
            (scenfit, repair_score, vec![])
        }
    };

    // Compute optimal repairs
    if repair_score > 0 {
        if let Some(max_repairs) = opt.max_repairs {
            let repairs = match opt.repair_mode {
                Some(RepairMode::OptGraph) if setting.ep => {
                    let mut repairs = vec![];
                    for new_edges in redges {
                        //TODO return only max_repairs solutions
                        let removes = get_opt_repairs_add_remove_edges_greedy(
                            &graph_facts,
                            &profiles,
                            &auto_inputs,
                            &new_edges,
                            scenfit,
                            repair_score,
                            max_repairs,
                            opt.threads,
                        )?;

                        for i in removes {
                            repairs.push(i);
                        }
                    }
                    repairs
                }
                Some(RepairMode::OptGraph) if !setting.ep => get_opt_repairs_add_remove_edges(
                    &graph_facts,
                    &profiles,
                    &auto_inputs,
                    scenfit,
                    repair_score,
                    max_repairs,
                    &setting,
                )?,
                Some(RepairMode::Flip) => get_opt_repairs_flip_edges(
                    &graph_facts,
                    &profiles,
                    &auto_inputs,
                    scenfit,
                    repair_score,
                    max_repairs,
                    &setting,
                )?,
                _ => get_opt_repairs_remove_edges(
                    &graph_facts,
                    &profiles,
                    &auto_inputs,
                    scenfit,
                    repair_score,
                    max_repairs,
                    &setting,
                )?,
            };

            if opt.json {
                let repairs: Vec<Vec<RepairOp>> = repairs
                    .iter()
                    .map(|set| {
                        set.iter()
                            .map(|symbol| into_repair(symbol).unwrap())
                            .collect()
                    })
                    .collect();

                let serialized = serde_json::to_string(&repairs)?;
                writeln!(out, ",\"Repair sets\":{serialized}")?;
            } else {
                for (c, r) in repairs.iter().enumerate() {
                    writeln!(out, "\n{}. Repair set:", c + 1)?;
                    for e in r {
                        let repair_op = into_repair(e)?;
                        writeln!(out, "  - {repair_op}")?;
                    }
                }
            }
        }
    }
    if opt.json {
        write!(out, "}}")?;
    }
    Ok(())
}

fn read_observations2(opt: &Opt) -> Result<Vec<(String, FactBase)>, anyhow::Error> {
    info!("Reading observations ...");
    let directory = fs::read_dir(&opt.observations_dir).context(format!(
        "unable to read directory '{}'",
        opt.observations_dir.display()
    ))?;
    let files: _ = directory.map_ok(|entry| {
        let observationfile = entry.path();
        let name = format!("{}", observationfile.display());

        match File::open(&observationfile) {
            Ok(file) => Ok((name, file)),
            Err(e) => Err(e).context(format!("Failed to open '{}'", name)),
        }
    });
    let profiles: _ = files.into_iter().map_ok(|res| match res {
        Ok((name, f)) => match profile_parser::read(&f, &name) {
            Ok(p) => Ok((name, p.to_facts())),
            Err(e) => Err(e).context(format!("Unable to parse '{}'", &name)),
        },
        Err(e) => Err(e),
    });
    let x: Result<Vec<(String, FactBase)>, anyhow::Error> = profiles.try_collect()?;
    x
}
fn read_observations(opt: &Opt, mut out: impl Write) -> Result<FactBase, anyhow::Error> {
    let profiles = read_observations2(opt)?;

    if opt.json {
        write!(out, ",\"observation files\":[")?;
    } else {
        writeln!(out, "\n## Observations\n")?;
    }

    let profiles_facts =
        profiles
            .iter()
            .fold(Ok(FactBase::new()), |mut acc, (name, profile)| {
                if let Inconsistent(reasons) = check_observations(&profile).unwrap() {
                    if acc.is_ok() {
                        acc = Err(InconsistenObsErr {
                            inconsistencies: vec![],
                        });
                    }
                    if let Err(mut i) = acc {
                        i.inconsistencies.push(InconsistentObs {
                            observations_file: name.clone(),
                            inconsistencies: reasons,
                        });
                        acc = Err(i);
                    }
                } else {
                    if let Ok(mut f) = acc {
                        f.union(&profile);
                        acc = Ok(f);
                    }
                }
                acc
            })?;

    let (names, _): (Vec<_>, Vec<_>) = profiles.into_iter().unzip();
    if opt.json {
        writeln!(out, "{}\n]", names.join(","))?;
    } else {
        for (c, name) in names.iter().enumerate() {
            writeln!(out, "{}. {name}", c + 1)?;
        }
    }
    Ok(profiles_facts)
}

fn compute_auto_inputs(
    opt: &Opt,
    graph_facts: &FactBase,
    out: &mut std::io::Stdout,
) -> Result<FactBase, anyhow::Error> {
    let auto_inputs = {
        if opt.auto_inputs {
            info!("Computing input nodes ...");
            let inputs = get_auto_inputs(graph_facts)?;
            let node_ids = get_node_ids_from_inputs(&inputs)?;
            if opt.json {
                write_auto_inputs_json(out, &node_ids)?;
            } else {
                write_auto_inputs_md(out, &node_ids)?
            }
            inputs
        } else {
            FactBase::new()
        }
    };
    Ok(auto_inputs)
}

fn get_setting(opt: &Opt) -> Setting {
    let setting = if opt.depmat {
        Setting {
            os: false,
            ep: true,
            fp: true,
            fc: true,
        }
    } else {
        Setting {
            os: true,
            ep: opt.elempath,
            fp: !opt.fwd_propagation_off,
            fc: !opt.founded_constraints_off,
        }
    };
    setting
}

fn read_network(opt: &Opt, mut out: impl Write) -> Result<Graph> {
    info!("Reading network model ...");
    if opt.json {
        writeln!(out, "\"Network file\":{:?}", opt.network_file)?;
    } else {
        writeln!(
            out,
            "\n## Network\n\n- Filename: {}",
            opt.network_file.display()
        )?;
    }
    let f = File::open(&opt.network_file)
        .context(format!("unable to open '{}'", opt.network_file.display()))?;
    let graph = cif_parser::read(&f)
        .context(format!("unable to parse '{}'", opt.network_file.display()))?;
    let network_statistics = graph.statistics();
    if opt.json {
        let serialized = serde_json::to_string(&network_statistics)?;
        writeln!(out, ",\"Network statistics\":{serialized}")?;
    } else {
        writeln!(out, "{network_statistics}")?;
    }
    Ok(graph)
}
